### ISSUE TITLE
[Feature request] Allow a sandboxed AppImage to launch other AppImages

### ISSUE THREAD

**Issue by Samueru-sama** [mgord9518/aisap#16]
Hello, I've recently came across [runimage](https://github.com/VHSgunzo/runimage) which is a portable container that uses bubblewrap and it is capable of launching appimages without issues unlike other similar solutions like conty. 

I contacted the developer and the reason it can do so is because it uses the `--cap-add ALL` flag with bubblewrap, but the issue of the appimages would be fixed by just adding the `CAP_SYS_ADMIN` capability. 

What do you think of doing this? Any downsides? If they are what about making that an option instead similar to the sandbox levels.

--------------------------------------------------------------

Also firefox web-browsers are not able to create their own sandbox under bwrap as [pointed out here](https://librewolf.net/installation/linux/#security) and enabling capabilities would allow for it to work under aisap.

**Comment by mgord9518** [comment]
It looks like a really cool project, do you know if bubblewrap can use that capability by default or does it need to be recompiled? 

As far as security, CAP_SYS_ADMIN is really powerful and can be a big attack vector. runimage doesn't seem to be security-oriented so it might not matter there but it definitely needs extra thought for aisap. Did the author mention what exactly allows for AppImages to launch using this capability? Does it allow launching SUID binaries or does it do something else for FUSE access?

But being that it takes bubblewrap permissions directly, it might actually be worth looking into as a "base OS" to launch AppImages through aisap on NixOS, Alpine, etc.

**Comment by Samueru-sama** [comment]
> It looks like a really cool project, do you know if bubblewrap can use that capability by default or does it need to be recompiled?

By what I was told, it is already build in bubblewrap and only the flag has to be passed.

> As far as security, CAP_SYS_ADMIN is really powerful and can be a big attack vector. runimage doesn't seem to be security-oriented so it might not matter there

I see, in that case what do you think of making it an option? iirc level 0 was removed from aisap so this could be added as the new "level 0" with a disclaimer that it isn't as safe as level 1, but for example the internal sandbox of firefox would work, and appimages could launch other appimages all while still keeping the rest of spooky locations isolated.

**Comment by mgord9518** [comment]
0 wasn't removed yet, but it's essentially useless and I do plan to remove it. Level 0 means run through bubblewrap but give all permissions, so it isn't actually sandboxed at all. My original idea was to have aisap use some sort of runtime to make it distro-agnostic, but I decided it was too much feature-creep and never added it. This seems to be exactly what runimage does though, so being able to piggyback off of it as an optional feature seems like a good plan.

I'll try to work on it this weekend, my current plan is to get the new aisap implementation working and then I can re-do the current one, greatly simplifying the codebase and adding these new features. Not sure how long it'll take though, I've been working on a few other projects that eat up most of my time

**Comment by mgord9518** [comment]
Just noticed I didn't even answer your question. Once I add runimage support, you should be able to launch AppImages from within other AppImages. Just like you suggested, it'll start as a minimum-security option, if we find it can be done securely then maybe it'll also be able to make it into levels 1, 2 and 3

**Comment by Samueru-sama** [comment]
> Just noticed I didn't even answer your question. Once I add runimage support, you should be able to launch AppImages from within other AppImages. Just like you suggested, it'll start as a minimum-security option, then if we find it can be done securely, maybe it'll also be able to make it into levels 1, 2 and 3

Note I don't current use runimage, just that I noticed that runimage can launch appimages despite using bubblewrap and doesn't have the same problem like conty or others where they can't launch appimages. 

In other words, passing the `--cap-add CAP_SYS_ADMIN` flag to bwrap in aisap should be all that's needed for sandboxed appimages to be able to launch other appimages. 

Sorry if this is what you meant, just that I'm bit confused by the phrase "runimage support". 

And I didn't know level 0 hasn't been removed, back in #14 I tried to launch aisap with level 0 and it did not work while it used to work before, so I assumed it was removed.

>Not sure how long it'll take though, I've been working on a few other projects that eat up most of my time

There is no hurry at all, I appreciate you still maintain this project despite no longer using appimages as much. If I were in your position this repo likely would have been archived lol.

**Comment by Samueru-sama** [comment]
Update: I was able to apply a similar fix to the Steam AppImage that we have, since it uses bubblewrap it had that issue that it could not launch appimages, which some steam games come as. 

However it needed more than just adding `--cap-add CAP_SYS_ADMIN` as that causes Steam to error out with no access to user namespaces. 

We had to patch bubblewrap [with this](https://github.com/Samueru-sama/Conty/blob/master/caps.patch) to fix that issue. 

I'm not sure how bad that patch is security wise.
